### ISSUE TITLE
`updated` and `patch` method on an infinite `Stream` should only force finite number of elements

### ISSUE THREAD

**Issue by Atry** [scala/bug#10791]
```  scala
Welcome to Scala 2.12.4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_152).
Type in expressions for evaluation. Or try :help.

scala> Stream.continually("*").updated(0, "new value") # Cause infinite loop
```

**Comment by Atry** [comment]
BTW, It seems the special `updated` also does not exist in https://github.com/scala/collection-strawman/blob/master/collections/src/main/scala/strawman/collection/immutable/LazyList.scala  

@szeiger

**Comment by SethTisue** [comment]
heh, nice catch

**Comment by joshlemer** [comment]
This seems to be fixed in 2.13.x branch:
```scala
    val x = Stream.continually("*").updated(0, "new value")
    println(x)
    println(x.take(10).toList)

    val y = Stream.continually("*").updated(4, "new value")
    println(y)
    println(y.take(10).toList)
```
prints
```
Stream(new value, ?)
List(new value, *, *, *, *, *, *, *, *, *)
Stream(*, ?)
List(*, *, *, *, new value, *, *, *, *, *)
```

**Comment by SethTisue** [comment]
@joshlemer could I interest you in adding test coverage on this, so we can be sure it doesn't regress?
